src/main/java/core/basesyntax/serviceimpl/ShopServiceImpl.java

```
@@ -0,0 +1,24 @@
+package core.basesyntax.serviceimpl;
+
+import core.basesyntax.common.FruitTransaction;
+import core.basesyntax.service.OperationStrategy;
+import core.basesyntax.service.ShopService;
+import core.basesyntax.strategy.OperationHandler;
+import java.util.List;
+
+public class ShopServiceImpl implements ShopService {
+    private final OperationStrategy operationStrategy;
+
+    public ShopServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        for (FruitTransaction transaction : transactions) {
```

You should also validate each `transaction` before use (inside the loop): ensure `transaction` is non-null and that `transaction.getOperation()` is non-null. If transactions may be malformed, either skip with a clear error or throw an informative RuntimeException. This aligns with the general input validation requirements (operation must be valid, non-null). Add checks before retrieving the handler to fail fast with informative messages.